Added more details to configurable tracing concept (#707)
kyma-project/community: concepts/observability-strategy/configurable-tracing/01-tracing-groundwork.md

# Configurable Tracing: Groundwork

## Current Situation and Motivation

![a](./assets/tracing-current.drawio.svg)

> The diagram shows that the current solution is based on a preconfigured Jaeger deployment which is serving as collector, in-memory storage and UI. The system components and the user workload are pushing the trace data directly to that Jaeger deployment using the Zipkin or Jaeger protocol. The setup does not support a neutral and unified way to integrate backends outside of the cluster at runtime. The storage is not durable and will not scale dependent on the load.

Reviewer: ```suggestion
The diagram shows that the current solution is based on a preconfigured Jaeger deployment, which serves as collector, in-memory storage, and UI. The system components and the user workload push the trace data directly to that Jaeger deployment using the Zipkin or Jaeger protocol. The current setup does not support a neutral and unified way to integrate backends outside of the cluster at runtime. The storage is not durable and will not scale dependent on the load.
```

As outlined in the [general strategy](../strategy.md), integration (and with that, changing the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can simply bring their own backends, if they already use a commercial offering or run their own infrastructure. To name just a few advantages, the data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, and kept away from any tampering or deletion attempts by hackers.

This concept proposes how to open up to those new scenarios by supporting convenient integration at runtime, leveraging vendor-neutral protocols and technologies.

## Requirements

### General
- Users need a way to outstream trace data into multiple external systems. There will be no production-ready solution provided by Kyma, so user need to integrate with something.
- Users need a way to add custom trace data from users' workload to that outstreams. Support for custom trace data is a major requirement for a distributed tracing solution.

> - Users need a way to influence the amount of traces being outstreamed. External systems will have a price dependent on load and traffic. Not relevant traces must be droppable with some strategies.

Reviewer: What does "with some strategies" mean? Can't we simply drop it?

```suggestion
- Users need a way to influence the amount of traces being outstreamed. External systems will have a price dependent on load and traffic. Not relevant traces must be droppable.
```

  - As a minimum, support one vendor-neutral output. It should be possible to chain your custom collector for specific conversions. For example, supporting the OLTP protocol will support most of the vendors already. Chaining a custom OpenTelemetry Collector can do custom conversion to a specific protocol.
  - Support typical settings for the supported outputs in a convenient way.
  - Non-typical settings should be possible as well, leaving the collector in an unsupported mode.
  - Secrets must be kept secret.
  - Secret bindings for the SAP BTP Operator must be supported, including secret rotations.
  - Scenarios must be isolated and have their own buffer management. If one backend is in a bad shape and cannot process any data anymore, data should still continue to be pushed to other backends.
  - Typical auth mechanisms for the integration must be supported, especially client-certificate-based solutions.
- Inputs
  - Inputs are predefined. In best case, users do not need to differentiate between inputs.
- Filter
  - Filtering of data (like dropping attributes or whole traces) must be possible.

### Pre-integration

> - Kyma system components are pre-integrated, so the predefined input in the collector serves them by default (eventing and serverless mainly)

Reviewer: ```suggestion
- Kyma system components are pre-integrated, so the predefined input in the collector serves them by default (Eventing and Serverless mainly).
```

Kyma components are [capitalized](https://kyma-project.io/community/guidelines/content-guidelines/02-style-and-terminology#capitalization).

> - Istio and serverless trace data is pre-integrated and can be easily filtered and de-selected.

Reviewer: ```suggestion
- Trace data from the Istio and Serverless components is pre-integrated and can be easily filtered and de-selected.
```
Mentioning "serverless" again - I assume this also refers to the Kyma component, so I'd capitalize it. But why mention it twice? If my assumption is wrong, please clarify what you mean.

### Ease of integration
- It should be easy to integrate trace data of a workload - with a push approach to a well-known internal URL in a vendor-neutral protocol
- Envoy trace data should be collected instantly (without user action).

### Local backend

> - Kyma will provide a blueprint based on Helm for installing a Jaeger based Deployment.

Reviewer: ```suggestion
- Kyma will provide a blueprint based on Helm for installing a Jaeger-based Deployment.
```
"Deployment" is about the Kubernetes term, right? Otherwise, don't capitalize it.
